abnfgen.1
.\" Copyright (c) 2002, 2006  Jutta Degener
.\"
.TH ABNFGEN 1  "4 March 2002"
.SH NAME
abnfgen \- ABNF-based test case generator
.SH SYNOPSIS
.na
.B abnfgen
[
.B \-hvclux_
] [
.B \-d
.I output-directory
]
.br
.ti +8
[
.B \-y
.I depth
]
[
.B \-n
.I ncases
]
.br
.ti +8
[
.B \-p
.I filename-pattern
]
[
.B \-w
.I prefix
]
[
.B \-r
.I seed
]
.br
.ti +8
[
.B \-s
.I start-symbol
]
[
.B \-t
.I tentative-file
]
.br
.ti +8
[
.I files...
]
.br
.ad
.SH DESCRIPTION
.LP
ABNF ("Advanced Backus Naur Form")
grammars are frequently used in RFCs to describe protocols or 
presentation formats for Internet Standards.
The \fIabnfgen\fP program produces text composed according to the rules of an
ABNF grammar, given the grammar.
Such text can then be used to test a parser that claims
to implement the grammar.
.SH OPTIONS
.TP
.B \-h
Print a brief usage.
.TP
.B \-v
Verbose mode.
Print a trace of the expanded grammar rules to standard error.
.TP
.B \-w 
Write seed.
Begin each generated file with \fIprefix\fR, followed by the seed
as a decimal number and a newline.
.TP
.B \-c
Try for complete coverage.
Rather than randomly picking productions, try to cover each
leaf and branch of the grammar.
.IP
For repetitions with * where maximum and minimum are less than
100 iterations apart, the repetition counts as fully covered
when both maximum and minimum have been produced.
(The stages in between are only tried once the extremes have
been covered.)
.IP
For repetitions with * where maximum and minimum are more than 
100 apart, only the minimum and some small number of repetitions
are tried; doubling the \fB-c\fP removes that limit and causes
the extremes to be tried even if the maximum is very large.
.IP
For character ranges with up to 256 elements, each elements
must be produced for full coverage.
For (Unicode) character ranges with more than 256 elements,
every possible value modulo 256 is produced at least once for
complete coverage.
.IP
A case-insensitive strings counts as fully covered once both
an all-uppercase and an all-lowercase version have been produced.
.TP
.B \-y
Descend freely through at most \fIdepth\fP nonterminal expansions
or repetitions.
After recursing that deeply, always pick that branch of the grammar that
terminates most quickly.
This can be used to limit expansion in grammars likely to
recurse infinitely such as \fBa = a a a a | "x"\fP.
The default depth is 100.
.TP
.B \-n
Rather than generating output on standard output (the default),
generate \fIncases\fP test files named ####.tst, where #### is 
replaced with the test's running number from 1 to \fIncases\fP. 
.TP
.B \-p
Name test cases using \fIfilename-pattern\fP.
A sequence of # characters in the pattern is replaced with
the running number of the test, padded to the specified size.
.TP
.B \-r
Initialize the random generator using \fIseed\fP.
For the same grammar and version of the software,
the same seed always creates the same subtree.
.TP
.B \-s
Start production with \fIstart-symbol\fP.
(Default: the first nonterminal defined in the grammar file.)
.TP
.B \-t
Read the nonterminals defined in \fItentative-file\fP
and use them if they don't get defined in the regular input files.
.TP
.B \-u
Reject any grammar that contains &lt;&gt;-enclosed prose.
.TP
.B \-x
Exclude the core set of definitions.
The RFC that defines the grammar also defines a core set of
symbols like CRLF, DIGIT, SP and so on.
Since version 0.9, abnfgen predefines these symbols as if
their definition had been passed in with -t.
Specify -x to suppress those predefinitions.
.TP
.B \-l
("legal") Disable extensions to RFC 4234: do not allow
case-sensitive literals in single quotes, do not allow
branch tagging with {}, and do not convert character constants
from Unicode to UTF-8.
.TP
.B \-_
Allow "_" as part of identifier names.
ABNF doesn't (it allows "-" instead), but many other
grammar systems do.
.SH GRAMMAR
The input grammar is a slight extension of ABNF with
provisions for literal strings and control of the chance
of descent into specific branches of the grammar.

.IP "\fInonterminal\fR \fB=\fR \fIexpression\fR"
The nonterminal expands to expression.
.IP "\fInonterminal\fR \fB=/\fR \fIexpression\fR"
Alternative to whatever else has been defined, the \fInonterminal\fP
can also expand to \fIexpression\fP.
.IP "\fIx\fP \fB/\fP \fIy\fP"
Either x or y.
.IP "\fIx\fP \fIy\fP"
X followed by y.
.IP \fB"abc"\fP
The case-insensitive string \fBabc\fR.  That is, one of 
ABC, ABc, AbC, Abc, aBC, aBc, abC, or abc.
There is no way of quoting " in a string; in a pinch, use <"> 
or %x22.
.IP "\fB'abc'\fR"
The case-sensitive string \fBabc\fR.
This is an extension to ABNF, and can be disabled using the \-l option.
In ABNF, 'abc' must be specified using the ascii values
of the characters, e.g. as %x61.62.63.
There is no way of quoting ' in a string; use '"'"'.
(That is, leave the single-quoted string, enter a double-quoted
string, the single quote, leave the double-quoted string, reenter
the single-quoted string.)
.IP \fB%\fIzN\fR
The character with base \fIz\fR value \fIN\fR.
Bases are: \fBx\fR (16),  \fBd\fR (10), and \fBb\fR (2).
.IP \fB%\fIzM\fR-\fIN\fR
Any character with values betwen \fIM\fR and \fIN\fR inclusive,
base \fIz\fR.
.IP "\fIM\fR\fB*\fR\fIN\fR \fIexpression\fR"
Between (inclusive) \fIM\fR and \fIN\fR repetitions
of \fIexpression\fR.
The default for \fIM\fR is 0, for \fIN\fR infinity.
.IP \fB[\fR\fIexpression\fR\fB]\fR
Same as \fB0*1(\fR\fIexpression\fR\fB)\fR, that is, an 
optional expression.
.IP (\fIexpression\fR)
Expands to \fIexpression\fR.
(Use parentheses for grouping.)
.SH "BRANCH CONTROL"
Each alternative in the grammar has a weight assigned to it.
Normally, these weights are all 1.
The higher the weight, the more likely it is that
a alternative is used when expanding the nonterminal
it occurs in.
Assign specific weights to an alternative by prefixing it
with \fB{\fR\fIN\fR\fB}\fR, where \fIN\fR is an integer.
For example
.br
.in +7
nt = {30} nt x / x
.br
x = {2} 'a' / {3} 'b'
.br
.in -7
should tend to produce output that is about 2/5th 'a',
about 3/5th 'b', and fairly lengthy.
.LP
Repetition can be branch controlled with two chance parameters
that govern whether to stop or continue on each step of the
repetition.
They're placed before the two counts around the *:
.br
.in +7
nt = {1}1*{3}10 'X'
.in -7
will tend to produce somewhere around 3 Xs on average.
This is an extension to RFC 4234 and can be disabled by
specifying \-l on the command line.
.SH BUGS
Please send problems, bugs, questions, desirable enhancements, etc. to:
.LP
.RS
jutta@pobox.com
.RE


ChangeLog
2004-10-10  Jutta Degener  <jutta@pobox.com>

	* Initial release.

2004-10-18  Jutta Degener  <jutta@pobox.com>	Version 0.2

	* Detect and complain about loops without alternatives in grammars.

2004-10-22  Jutta Degener  <jutta@pobox.com>    Version 0.3

   	* -v now means "dump a rule trace to stderr as you expand".

	* -w now accepts a string as an argument and prints the seed prefixed
		with that argument (e.g., "# ") in each output file.

	* The version is now printed as part of the usage string.

	* Fewer alternatives and concatenations with only one element.

2004-10-22  Jutta Degener  <jutta@pobox.com>    Version 0.4
	
	* Don't complain that "a" in a = b c b  doesn't terminate.

2004-12-12  Jutta Degener  <jutta@pobox.com>    Version 0.5
	
	* 0.4 README referred to version 0.1, listed wrong URL.

2005-02-02  Jutta Degener  <jutta@pobox.com>    Version 0.6
	
	* UTF-8 encoder's 6-byte prefix must be 0xFC, not 0xF8.

2006-10-18  Jutta Degener  <jutta@pobox.com>    Version 0.7
	
	* Fixed off-by-one allocation error in the STRMALCPY() macro.
		Thanks to Philip Gunther for pointing it out!

2006-10-26  Jutta Degener  <jutta@pobox.com>    Version 0.8
	
	* Don't go into endless recursion over nested alternatives.
		Thanks to Stéphane Bortzmeyer, author of "Euthatius",
		the Haskell abnf generator, for pointing out the bug!

2006-11-13  Jutta Degener  <jutta@pobox.com>	Version 0.9

	* Don't overload builtin round().
	* Don't allow spaces in N*Mexpr.
	    	(Thanks again, Stéphane!)

2006-11-23  Jutta Degener  <jutta@pobox.com>	Version 0.10
	
	* 4FOO now performs 4 expansions of FOO, not just one.
	  (Thanks to Lisa Dusseault for pointing this out in
	  the course of her Implementation Report.)
	* Invert the meaning of -u, and document its existence.


2006-12-18  Jutta Degener  <jutta@pobox.com>	Version 0.11

	* Don't allow "_" in identifier names.  (Thanks, Stéphane!)
	  The new "-_" command line option re-enables the
	  previous behavior.

2006-12-21  Jutta Degener  <jutta@pobox.com>	Version 0.12

	* Don't let %d12-%d13 slip through (the correct syntax
	  is %d12-13).

2007-01-16  Jutta Degener  <jutta@pobox.com>	Version 0.13

	* When evaluating [], don't by default pick the non-empty
	  version (Thanks, Lisa!)

2007-04-10  Jutta Degener <jutta@pobox.com>	Version 0.14

	* Don't stop recursive checking when encountering zero repetitions.
	    	(Thanks again, Stéphane!)

2007-07-10  Jutta Degener <jutta@pobox.com>	Version 0.15

	* Handle EOF in strings, octet strings, and <>-enclosed prose.

2007-12-05  Jutta Degener <jutta@pobox.com>	Version 0.16

	* New option -l turns off extensions.



abnfgen.c
/*
 * $Id: abnfgen.c,v 1.5 2004/10/23 01:21:52 jutta Exp $
 * $Log: abnfgen.c,v $
 * Revision 1.5  2004/10/23 01:21:52  jutta
 * keep track of which nonterminals we've complained about.
 *
 * Revision 1.4  2004/10/23 00:50:32  jutta
 * add verbose expansions
 *
 * Revision 1.3  2004/10/19 04:45:55  jutta
 * reset the default depth to 100.
 *
 * Revision 1.2  2004/10/18 07:38:10  jutta
 * Bump default depth to a 1000.
 *
 * Revision 1.1  2004/10/11 00:36:50  jutta
 * Initial revision
 *
 *
 */

#include <getopt.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <stdarg.h>
#include <time.h>
#include <errno.h>

#include "abnfgenp.h"

static char const rcsid[]
	= "$Id: abnfgen.c,v 1.5 2004/10/23 01:21:52 jutta Exp $";

static char const RFC_4234_CORE[] =
	"ALPHA	=  %x41-5A / %x61-7A   ; A-Z / a-z\n"
	"BIT   	=  \"0\" / \"1\"\n"
	"CHAR  	=  %x01-7F	; any 7-bit US-ASCII character,\n"
	"			;  excluding NUL\n"
	"CR    	=  %x0D		; carriage return\n"
	"CRLF  	=  CR LF	; Internet standard newline\n"
	"CTL   	=  %x00-1F / %x7F\n"
	"			; controls\n"
	"DIGIT 	=  %x30-39	; 0-9\n"
	"DQUOTE	=  %x22		; \" (Double Quote)\n"
	"HEXDIG	=  DIGIT / \"A\" / \"B\" / \"C\" / \"D\" / \"E\" / \"F\"\n"
	"HTAB  	=  %x09		; horizontal tab\n"
	"LF    	=  %x0A		; linefeed\n"
	"LWSP  	=  *(WSP / CRLF WSP)\n"
	"			; linear white space (past newline)\n"
	"OCTET 	=  %x00-FF	; 8 bits of data\n"
	"SP    	=  %x20\n"
	"VCHAR 	=  %x21-7E 	; visible (printing) characters\n"
	"WSP   	=  SP / HTAB	; white space\n";

static void usage(char const * progname)
{
	fprintf(stderr, "%s version %s; send bugs to <jutta@pobox.com>\n"
		"Usage: %s [options][inputs]\n", progname, VERSION, progname);
	fputs(
	"Output:\n"
	"   -d dir	   -- write output files to directory <dir>\n"
	"   -n n	   -- write <n> output files with #=1..n\n"
	"   -p pat#.suf    -- create filenames \"pat#.suf\"\n",
		stderr);
	fputs(
	"Processing:\n"
	"   -c 	           -- attempt full coverage\n"
	"   -l 	           -- (\"legal\") strict RFC 4234 only\n"
	"   -y n	   -- limit recursion depth to <n>\n"
	"   -r n           -- seed random generator with <n>\n"
	"   -s nonterminal -- start symbol is <nonterminal>\n"
	"   -t file        -- include tentative definitions in <file>\n"
	"   -u             -- reject rules that contain <prose>\n",
	   stderr);
	fputs(
	"Miscellaneous:\n"
	"   -x 	           -- exclude (don't preload) the core syntax\n"
	"   -h 	           -- print this statement\n"
	"   -w prefix      -- write seed to stdout, prefixed with <prefix>\n"
	"   -v 	           -- verbose; write rule trace to stderr\n",
		stderr);
	exit(64);
}

void ag_error(ag_handle * ag, char const * fmt, ...)
{
	va_list ap;

	if (ag->errors++ >= 15) {
		fprintf(stderr, "Too many errors; abort.\n");
		exit(0);
	}
	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
}

void * ag_emalloc(
	ag_handle * ag,
	char const * purpose,
	size_t 	     n)
{
	char * tmp = malloc(n);
	if (!tmp) {
		ag_error(ag,
			"failed to allocate %lu bytes for %s\n",
			(unsigned long)n, purpose);
		return 0;
	}
	return tmp;
}

static int open_output_file(
	ag_handle * ag,
	char const * dir,
	char const * pat,
	int 	     i,
	FILE      ** f_out,
	char      ** name_out)
{
	char const * r;
	char *w, * tmp = ag_emalloc(ag, "output file name",
		strlen(dir) + 1 + strlen(pat) + 42);

	if (!tmp) return -1;

	w = tmp;
	if (dir) {
		while (*w = *dir++)
			w++;
		if (w > tmp && w[-1] != '/')
			*w++ = '/';
	}
	for (r = pat; *r;)
		if (*r != '#')
			*w++ = *r++;
		else {
			int n = 0;
			while (*r == '#')
				r++, n++;
			sprintf(w, "%.*u", n, i);
			w += strlen(w);
		}
	*w = 0;

	if (!(*f_out = fopen(tmp, "w"))) {
		ag_error(ag, "%s: failed to open %s for output: %s\n",
			ag->progname, tmp, strerror(errno));
		(void)free(tmp);
		return -1;
	}
	*name_out = tmp;
	return 0;
}

static int ag_process(
	char const  	   * progname,
	int 	      	     verbose,
	int 	      	     understand_prose,
	int 	      	     legal,
	int 	      	     underscore_in_identifiers,
	int 	      	     full_coverage,
	char const  	   * output_directory,
	char const  	   * pattern,
	unsigned int	     seed,
	char const	   * seed_prefix,
	unsigned int	     n_cases,
	unsigned int	     depth,
	char const	   * start_symbol,
	char const * const * av_tentative,
	char const * const * av)
{
	ag_handle    ag[1];
	unsigned int  i;

	memset(ag, 0, sizeof(*ag));
	ag->verbose          = verbose;
	ag->full_coverage    = full_coverage;
	ag->progname         = progname;
	ag->seed             = seed;
	ag->seed_prefix      = seed_prefix;
	ag->understand_prose = !!understand_prose;
	ag->legal	     = !!legal;
	ag->underscore_in_identifiers = !!underscore_in_identifiers;

	hinit(&ag->symbols, 	 char, 1024 * 8);
	hinit(&ag->nonterminals, ag_nonterminal, 512);
	hinit(&ag->complained,   char, 512);

	if (start_symbol)
		ag->start_symbol = ag_symbol_make_lowercase(ag,
			start_symbol);

	if (av_tentative)
		for (i = 0; av_tentative[i]; i++) {

			if (*av_tentative[i] == '\0')
				ag_input_string(ag, RFC_4234_CORE,
				  "ABNF core grammar included in RFC 4234",
				  1);
			else
			{
				FILE * f = fopen(av_tentative[i], "r");
				if (!f) {
					ag_error(ag,
						"%s: can't open %s for input: "
						"%s\n", ag->progname, 
						av_tentative[i],
						strerror(errno));
				}
				else {
					ag_input(ag, f, av_tentative[i], 1);
					fclose(f);
				}
			}
		}

	if (!*av)
		ag_input(ag, stdin, "*standard input*", 0);
	else {
		for (i = 0; av[i]; i++) {
			FILE * f = fopen(av[i], "r");
			if (!f) {
				ag_error(ag, "%s: can't open %s for input: "
					"%s\n", ag->progname,
					av[i], strerror(errno));
			}
			else {
				ag_input(ag, f, av[i], 0);
				fclose(f);
			}
		}
	}

	ag_check(ag);
	if (n_cases && !ag->errors) {

		if (output_directory) {
			
			if (  mkdir(output_directory, 0777) < 0
			   && errno != EEXIST) {

				ag_error(ag,
				 "%s: can't create directory %s: %s\n",
					ag->progname, output_directory,
					strerror(errno));
				return 1;
			}
			for (i = 0; i < n_cases && !ag->errors; i++) {

				char * out_name;
				FILE * out_file;

				if (open_output_file(ag,
					output_directory, pattern, i + 1,
					&out_file, &out_name))

					break;

				ag_output(ag, out_file, out_name, i + 1, depth);
				fclose(out_file);
				(void)free(out_name);
			}
		}
		else
			for (i = 1; i <= n_cases && !ag->errors; i++)  {
				ag_output(ag, stdout, "*standard output*", i,
					depth);
			}
	}
	return ag->errors ? 1 : 0;
}

int main(int ac, char ** av)
{
	unsigned int	  seed = (unsigned int)time(0) * (unsigned int)getpid(),
			  n_cases = 1, verbose = 0, depth = 100;
	char		* seed_prefix = NULL;
	char		* output_directory = 0;
	char const	* pattern = "####.tst";
	char const	* progname;
	int 		  opt, result;
	int		  full_coverage = 0;
	int		  explicit_seed = 0;
	int		  understand_prose = 1;
	int		  legal = 0;
	int		  exclude_core = 0;
	int		  underscore_in_identifiers = 0;
	extern char 	* optarg;
	extern int	  optind;
	char 	       ** av_tentative = 0;
	char		* start_symbol = 0;

	if (progname = strrchr(av[0], '/')) progname++;
	else progname = av[0];

	av_tentative = argvadd(av_tentative, "");

	while ((opt = getopt(ac, av, "cd:hln:p:r:s:t:uvw:x_y:")) != EOF)
		switch (opt) {
		case '_':
			underscore_in_identifiers++;
			break;

		case 'c':
			full_coverage++;
			break;

		case 'd':
			output_directory = optarg;
			break;

		case 'y':
			if (sscanf(optarg, "%u", &depth) != 1) {
				fprintf(stderr, "%s: expected non-negative "
					"integer with -y, got \"%s\"\n",
					progname, optarg);
				usage(progname);
			}
			break;

		case 'n':
			if (sscanf(optarg, "%u", &n_cases) != 1) {
				fprintf(stderr, "%s: expected non-negative "
					"integer with -n, got \"%s\"\n",
					progname, optarg);
				usage(progname);
			}
			break;

		case 'p':
			pattern = optarg;
			break;

		case 'r':
			explicit_seed = 1;
			if (sscanf(optarg, "%u", &seed) != 1) {
				fprintf(stderr, "%s: expected non-negative "
					"integer with -r, got \"%s\"\n",
					progname, optarg);
				usage(progname);
			}
			break;

		case 's':
			start_symbol = optarg;
			break;

		case 't':
			av_tentative = argvadd(av_tentative, optarg);
			break;

		case 'u':
			understand_prose = 0;
			break;

		case 'v':
			verbose++;
			break;

		case 'w':
			explicit_seed = 1;
			seed_prefix = optarg;
			break;

		case 'x':
			exclude_core = 1;
			break;

		case 'l':
			legal = 1;
			break;

		case 'h':
		default:
			usage(progname);
		}

	if (full_coverage && explicit_seed && n_cases > 1) {
		fprintf(stderr, "%s: warning: seeds for other than the first "
			"run of \"%s -c\" won't yield the original result.\n",
			progname, progname);
	}
	result = ag_process(progname, verbose, understand_prose, legal,
		underscore_in_identifiers, full_coverage,
		output_directory, pattern, seed, seed_prefix, n_cases, depth,
		start_symbol,
		(char const * const *)av_tentative + exclude_core,
		(char const * const *)av + optind);
	argvfree(av_tentative);
	return result;
}
